File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    int[] numbers;
    int target;
    int answer;
    
    void dfs(int index, int sum){
        
        if (index == numbers.length){
            if (sum == target) { answer ++;
                return; // 탈출
            }
        }
        
        // index가 전체길이 보다 커지면 return
        if (index >= numbers.length) {
            return;
        }
        
        //수행동작 
        dfs(index + 1 , sum + numbers[index]);
        dfs(index + 1, sum - numbers[index]);
    }
    
    public int solution(int[] numbers, int target) {
        answer = 0;
        this.numbers = numbers;
        this.target = target;      
        dfs(0,0); 
        return answer;
    }

}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def solution(numbers, target):
    n = len(numbers)
    answer = 0
    
    #카운트 하나씩 해준다
    def dfs(cnt, result):
        if cnt == n: # 카운트가 n과 같?
            if result == target: # 결과값이 같?
                nonlocal answer 
                answer += 1 # answer +1
            return #아니라면 return 
        else:  # 카운트가 n과 같지 않다면
            dfs(cnt+1, result+numbers[cnt])
            dfs(cnt+1, result-numbers[cnt])
    dfs(0,0)
    return answer